### ISSUE TITLE
[Bug] 3rdparty/tvm/src/runtime/relax_vm/lm_support.cc:421: InternalError: Check failed: sampled_index >= 0 (-1 vs. 0) :

### ISSUE THREAD

**Issue by saum7800** [mlc-ai/mlc-llm#1279]
## 🐛 Bug

Getting "3rdparty/tvm/src/runtime/relax_vm/lm_support.cc:421: InternalError: Check failed: sampled_index >= 0 (-1 vs. 0) :" when trying to run a custom model on an android phone. My custom model is essentially exactly the same as RedPajama-INCITE-chat-3b-v1 with certain weights zeroed out due to pruning. you can find it here: https://huggingface.co/ssgandhi7800/Linear_pruned_RedPajama3B-q4f16_1

## To Reproduce

1. Follow the compilation guide with target as android for hf_path `Advaith28/Linear_pruned_RedPajama3B-q4f16_1` and `--use-safetensors`
2. Follow the android guide by adding `ssgandhi7800/Linear_pruned_RedPajama3B-q4f16_1` to app-config.json
3. Run the app on an android phone

The app downloads and runs, but ends up with this error: `mlc-llm/3rdparty/tvm/src/runtime/relax_vm/lm_support.cc:421: InternalError: Check failed: sampled_index >= 0 (-1 vs. 0) : `

## Expected behavior

It should run normally, how RedPajama runs

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): Android
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): Ubuntu
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...): Samsung
 - How you installed MLC-LLM (`conda`, source): conda
 - How you installed TVM-Unity (`pip`, source):
 - Python version (e.g. 3.10): 3.10
 - GPU driver version (if applicable):
 - CUDA/cuDNN version (if applicable):
 - TVM Unity Hash Tag (`python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))"`, applicable if you compile models):
 - Any other relevant information:

## Additional context

I updated my mlc-chat-config to mimic the RedPajama one too. I am not sure why my entire process would be any different from using the normal RedPajama model.

**Comment by Hzfengsy** [comment]
This is due to `nan` output. BTW, I improved the error msg in https://github.com/apache/tvm/pull/16141

**Comment by saum7800** [comment]
Oh I see. That's interesting! Do you know any workarounds I can make? Say maybe some way where all NaNs can be converted to/treated as zeros? Or something else?

Thanks for making a PR to clarify the error message for future releases!

**Comment by Hzfengsy** [comment]
I think there is no workarounds. `NaN` means the model can not output meaningful results. Also, NaNs are not equal to zeros. So my suggestion is to change the model/weight or change the quantization method (e.g. to no quant, q0f16)

The initial problem of this issue is solved, so I will close it. If you have further questions about quantization and model selection, feel free to open a new one :)

**Comment by junrushao** [comment]
Quick question - just wanted to double check: does it always give NaN or just occasionally?

**Comment by saum7800** [comment]
It was always giving NaN. I guess either:
1. during the qunatization of weights, the zeroed out weights became NaN.
2. activations upon being multiplied with 0 become NaN.

**Comment by junrushao** [comment]
This is definitely weird. Do you think

**Comment by saum7800** [comment]
I was able to reproduce the issue on my Ubuntu Linux (compiled for Vulkan).

Since then, I realised that the model I was compiling from had been stored in int8 format whereas MLC probably expects fp16/32 weights from the input model, which must have created issues for the quantization and led to the NaNs error? Maybe it would have been helpful to error out during compilation stating that these weights are int8 and cannot be quantized.

I updated our model to be a fp16 model and performed all the steps again and everything is working fine now. Thank you for the help and this truly wonderful library @junrushao  and @Hzfengsy !
